# HG changeset patch
# User Magnus Melin <mkmelin+mozilla@iki.fi>
# Date 1695709242 -10800
# Node ID 88770ab5141077c8f0ebce98f6d4209ad33d6a3d
# Parent  97b5c8f94797da4c0a5a10d83ba1ce8818c127b6
Bug 1854064 - Don't load accounts that don't have a server hostname. r=john.bieling

The UI doesn't allow such accounts. If somehow we got into the state of having no hostname,
the account is doomed and can never do anything useful, so don't load it at all.
If we allow such accounts, they will blow up the UI since we then end up with folders
that don't know which server they are belong to.

Differential Revision: https://phabricator.services.mozilla.com/D188957

diff --git a/mailnews/base/src/nsMsgAccount.cpp b/mailnews/base/src/nsMsgAccount.cpp
--- a/mailnews/base/src/nsMsgAccount.cpp
+++ b/mailnews/base/src/nsMsgAccount.cpp
@@ -6,17 +6,17 @@
 #include "prprf.h"
 #include "plstr.h"
 #include "prmem.h"
 #include "nsIComponentManager.h"
 #include "nsIServiceManager.h"
 #include "nsCRTGlue.h"
 #include "nsCOMPtr.h"
 #include "nsIMsgFolderNotificationService.h"
-
+#include "nsPrintfCString.h"
 #include "nsIPrefService.h"
 #include "nsIPrefBranch.h"
 #include "nsMsgAccount.h"
 #include "nsIMsgAccount.h"
 #include "nsIMsgAccountManager.h"
 #include "nsIObserverService.h"
 #include "mozilla/Services.h"
 #include "nsServiceManagerUtils.h"
@@ -87,16 +87,25 @@ nsresult nsMsgAccount::createIncomingSer
   nsCOMPtr<nsIMsgAccountManager> accountManager =
       do_GetService("@mozilla.org/messenger/account-manager;1", &rv);
   NS_ENSURE_SUCCESS(rv, rv);
 
   nsCOMPtr<nsIMsgIncomingServer> server;
   rv = accountManager->GetIncomingServer(serverKey, getter_AddRefs(server));
   NS_ENSURE_SUCCESS(rv, rv);
 
+  nsCString hostname;
+  rv = server->GetHostName(hostname);
+  NS_ENSURE_SUCCESS(rv, rv);
+  if (hostname.IsEmpty()) {
+    NS_WARNING(
+        nsPrintfCString("Server had no hostname; key=%s", serverKey.get()).get());
+    return NS_ERROR_UNEXPECTED;
+  }
+
   // store the server in this structure
   m_incomingServer = server;
   accountManager->NotifyServerLoaded(server);
 
   return NS_OK;
 }
 
 NS_IMETHODIMP
